storm-core/src/jvm/backtype/storm/scheduler/resource/ResourceAwareScheduler.java

```diff
@@ -28,94 +32,272 @@
+                            this.conf.get(Config.RESOURCE_AWARE_SCHEDULER_PRIORITY_STRATEGY), e.getMessage());
+                    break;
+                }
+            }
+            TopologyDetails td = null;
+            try {
+                //need to re prepare since scheduling state might have been restored
+                schedulingPrioritystrategy.prepare(this.topologies, this.cluster, this.userMap, this.nodes);
+                //Call scheduling priority strategy
+                td = schedulingPrioritystrategy.getNextTopologyToSchedule();
+            } catch (Exception e) {
+                LOG.error("Exception thrown when running priority strategy {}. No topologies will be scheduled! Error: {} StackTrack: {}"
+                        , schedulingPrioritystrategy.getClass().getName(), e.getMessage(), Arrays.toString(e.getStackTrace()));
+                break;
+            }
+            if (td == null) {
+                break;
+            }
+            scheduleTopology(td);
+        }
+    }
+
+    public void scheduleTopology(TopologyDetails td) {
+        User topologySubmitter = this.userMap.get(td.getTopologySubmitter());
+        if (cluster.getUnassignedExecutors(td).size() > 0) {
+            LOG.debug("/********Scheduling topology {} from User {}************/", td.getName(), topologySubmitter);
+
+            SchedulingState schedulingState = this.checkpointSchedulingState();
+            IStrategy RAStrategy = null;
+            try {
+                RAStrategy = (IStrategy) Utils.newInstance((String) td.getConf().get(Config.TOPOLOGY_SCHEDULER_STRATEGY));
+            } catch (RuntimeException e) {
+                LOG.error("failed to create instance of IStrategy: {} with error: {}! Topology {} will not be scheduled.",
+                        td.getName(), td.getConf().get(Config.TOPOLOGY_SCHEDULER_STRATEGY), e.getMessage());
+                this.restoreCheckpointSchedulingState(schedulingState);
+                //since state is restored need the update User topologySubmitter to the new User object in userMap
+                topologySubmitter = this.userMap.get(td.getTopologySubmitter());
+                topologySubmitter.moveTopoFromPendingToInvalid(td);
+                this.cluster.setStatus(td.getId(), "Unsuccessful in scheduling - failed to create instance of topology strategy "
+                        + td.getConf().get(Config.TOPOLOGY_SCHEDULER_STRATEGY) + ". Please check logs for details");
+                return;
+            }
+            IEvictionStrategy evictionStrategy = null;
+            while (true) {
+                SchedulingResult result = null;
+                try {
+                    //Need to re prepare scheduling strategy with cluster and topologies in case scheduling state was restored
+                    RAStrategy.prepare(this.topologies, this.cluster, this.userMap, this.nodes);
+                    result = RAStrategy.schedule(td);
```

Comment:
well we need to create a IStrategy object each time since topologies can have different strategies.  Even multiple topologies use the same strategy, we should still need to isolate the scheduling of each topology.  We can cache an IStrategy object but we would then need to have a way to reset all the data structures in the IStrategy object will may contain a lot of user defined data structures thus, I think we should just recreate the IStrategy object